Source/ObjectiveDropboxOfficial/Shared/Handwritten/Networking/DBGlobalErrorResponseHandler.m

```diff
@@ -0,0 +1,171 @@
+///
+/// Copyright (c) 2016 Dropbox, Inc. All rights reserved.
+///
+
+#import "DBGlobalErrorResponseHandler.h"
+#import <objc/runtime.h>
+
+static DBNetworkErrorResponseBlock s_networkErrorResponseBlock = nil;
+static NSOperationQueue *s_networkErrorQueue;
+
+static NSMutableDictionary<Class, DBRouteErrorResponseBlock> * _Nullable s_routeErrorToResponseBlock;
+static NSMutableDictionary<Class, NSOperationQueue *> * _Nullable s_routeErrorToQueue;
+
+static NSLock *s_lock;
+
+@implementation DBGlobalErrorResponseHandler
+
++ (void)initialize {
+  static dispatch_once_t once;
+  dispatch_once(&once, ^{
+    s_networkErrorQueue = [NSOperationQueue mainQueue];
+
+    s_routeErrorToResponseBlock = [NSMutableDictionary new];
+    s_routeErrorToQueue = [NSMutableDictionary new];
+
+    s_lock = [NSLock new];
+  });
+}
+
++ (void)registerRouteErrorResponseBlock:(DBRouteErrorResponseBlock)routeResponseBlock
+                         routeErrorType:(id)routeErrorType {
+  [self registerRouteErrorResponseBlock:routeResponseBlock routeErrorType:routeErrorType queue:nil];
+}
+
++ (void)registerRouteErrorResponseBlock:(DBRouteErrorResponseBlock)routeResponseBlock
+                         routeErrorType:(id)routeErrorType
+                                  queue:(NSOperationQueue *)queue {
+  NSOperationQueue *queueToUse = queue ?: [NSOperationQueue mainQueue];
+
+  [s_lock lock];
+  s_routeErrorToResponseBlock[routeErrorType] = routeResponseBlock;
+  s_routeErrorToQueue[routeErrorType] = queueToUse;
+  [s_lock unlock];
+}
+
++ (void)removeRouteErrorResponseBlockWithRouteErrorType:(id)routeErrorType {
+  [s_lock lock];
+  [s_routeErrorToResponseBlock removeObjectForKey:routeErrorType];
+  [s_lock unlock];
+}
+
++ (void)registerNetworkErrorResponseBlock:(DBNetworkErrorResponseBlock)networkErrorResponseBlock {
+  [self registerNetworkErrorResponseBlock:networkErrorResponseBlock queue:nil];
+}
+
++ (void)registerNetworkErrorResponseBlock:(DBNetworkErrorResponseBlock)networkErrorResponseBlock
+                                    queue:(NSOperationQueue * _Nullable)queue {
+  NSOperationQueue *queueToUse = queue;
+
+  [s_lock lock];
+  s_networkErrorResponseBlock = networkErrorResponseBlock;
+
+  if (queueToUse) {
+    s_networkErrorQueue = queueToUse;
```

Comment:
This is kind of an edge case, but if you set one network error block on some queue, then set another error block with no queue provided, it'll still use the old queue.